Netcode: Server-side order sync and optimizations to reduce latency

adammitchelldev opened the pull request:
An updated version of #17900. See that PR for a more in-depth discussion of the changes.

This code change improves input responsiveness in multiplayer games to be as small as possible (from well over 300ms). Please give it a spin and try a comparison for yourself.

It's expected that there will be a few instabilities: replays and singleplayer may be broken.

## Install and Use

Dedicated servers must launch with the additional env option: `UseNewNetcode="True"`. Game clients can host games with the new netcode using the checkbox provided in the Settings->Advanced tab.

You need to ensure that you have a valid version tag in the game to play online with the server browser, which can be done reliably with the following commands:

On Windows with PowerShell use:
```
.\make.cmd version
.\make.cmd all
.\launch-game.cmd Game.Mod=ra
```

On Linux:
```
make version
make all
./launch-game.cmd Game.Mod=ra
```

[Script for a dedicated server on Ubuntu 20.04 (needs specific version for Mono)](https://gist.github.com/adammitchelldev/4986da1e6d65d79ed23db64e07afbfdd)

You can try executing it directly with:
```
curl -sSf https://gist.githubusercontent.com/adammitchelldev/4986da1e6d65d79ed23db64e07afbfdd/raw/634e6899842cd41c34dba8c34c89c453ce4cfb9b/netcode.sh | bash
```

Leonardo-Ferreira commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -146,13 +148,14 @@ public virtual void SendSync(int frame, byte[] syncData)
 			ms.WriteArray(BitConverter.GetBytes(frame));
 			ms.WriteArray(syncData);
-			Send(ms.GetBuffer());
+			Send(ms.ToArray());
```
`ToArray` is a costly operation (https://github.com/microsoft/referencesource/blob/master/mscorlib/system/io/memorystream.cs line 550) and should be used only if needed. If `ms.Capacity == ms.Length` you should use `GetBuffer`

Leonardo-Ferreira commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,6 +335,30 @@ void NetworkConnectionReceive()
+			reader.ReadByte();
+			var framesToAck = reader.ReadInt16();
+
+			var ms = new MemoryStream();
```
Avoid instantiating a empty memoryStream... the first write on it will secure 256 bytes of storage, which I think is not what you want...

Leonardo-Ferreira commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,6 +335,30 @@ void NetworkConnectionReceive()
+			}
+
+			AddPacket(new ReceivedPacket { FromClient = LocalClientId, Data = ms.ToArray() });
+		}
```
Dispose the stream you instantiated above

Leonardo-Ferreira commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -333,27 +367,54 @@ public override void SendSync(int frame, byte[] syncData)
 			ms.WriteArray(syncData);
-			queuedSyncPackets.Add(ms.GetBuffer());
+
+			queuedSyncPackets.Add(ms.ToArray()); // TODO: re-add sync packets
```
check comment above about ToArray

Leonardo-Ferreira commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -333,27 +367,54 @@ public override void SendSync(int frame, byte[] syncData)
+			if (ordersArray.Length > 0)
+			{
+				// Write our packet to be acked
+				var ackMs = new MemoryStream();
```
Dispose at the end

OpenOP2Admin commented:
If this is what you say it is, please implement this you will do a great service

deleted-user-1 commented:
How do we get this tested and going? Can someone make a build ("doubleclick-ready") for the discord people and announce this there and host a server with these changes to get testing going? This has potential to remove a substantial part of the ever-present multiplayer lag, everyone should be hyped! Are downstream people interested or already using this? Polite ping to @netnazgul and @jrb0001.

adammitchelldev commented:
Nice work matjaeck!

On Thu, 7 Jan 2021, 16:09 matjaeck, <notifications@github.com> wrote:

> How do we get this tested and going? Can someone make a build
> ("doubleclick-ready") for the discord people and announce this there and
> host a server with these changes to get testing going? This has potential
> to remove a substantial part of the ever-present multiplayer lag, everyone
> should be hyped! Are downstream people interested or already using this?
> Polite ping to @netnazgul <https://github.com/netnazgul> and @jrb0001
> <https://github.com/jrb0001>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/18783#issuecomment-756211923>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOKXUH3A27HSFRUXV2STFDSYXMCDANCNFSM4TC5KOYQ>
> .
>

deleted-user-1 commented:
It obviously won't work without attention from the maintainers and code review but my hope is that feedback from somewhat coordinated testing in bigger games (which should be possible) can help to make some progress here, even if there are other priorities like the remasters and .NET5 support and many other PRs blocking it.

jrb0001 commented on OpenRA.Game/Network/Order.cs:
```diff
@@ -21,7 +21,8 @@ public enum OrderType : byte
 		Handshake = 0xFE,
-		Fields = 0xFF
+		Fields = 0xFF,
+		Ack = 0x10
```
This will need a bump in the protocol version.

adammitchelldev replied:
Now that there's a lobby option `UseNewNetcode` it shouldn't, I'd expect this to be backward compatible with existing servers.

pchote replied:
Alternative servers may not be able to process that option, so bumping the orders protocol version is the right thing to do.

jrb0001 commented on OpenRA.Game/Server/ProtocolVersion.cs:
```diff
@@ -36,6 +36,11 @@ public static class ProtocolVersion
+		// When the frame of a packet is 0, it is an immediate order, and may or may not be relayed to clients e.g. chat
+		//
+		// When the frame is not 0, it will always be relayed to all clients
+		// and a blank packet sent to the sender to act as an ack (used by adaptive order latency)
```
This ack system makes everything considerably more complex and doesn't give any latency advantage. The slightly higher bandwidth usage also shouldn't really make a difference. So why don't you simply send everything through the server without that local loopback special case?

adammitchelldev replied:
That's how I started. I added the ack system on top once it worked without it. It doesn't actually add that much complexity on top of the original code, I'd say most of the complexity is in the server corralling of orders.

Hopefully, we replace the network protocol with UDP as the next step. Despite the `NoDelay` option, TCP still has other issues that can cause unnecessary latency under small amounts of packet loss. There are far more effective ack and bandwidth control schemes for real-time communication that can be written with UDP and will work under very poor conditions. The logic for this stuff will be a lot more complex than the simple loopback acks here, but it will likely supersede it and become an all-in-one protocol that just straight up replaces the current protocol and deciphering logic.

However, this has got me thinking some, so I've got my favourite calculations napkin. This is going to be for the current OpenRA, not this branch: normal game ticks at 40ms and net ticks at 120ms (every 3 frames). Let's assume we have an 8 player game, just for fun.

OpenRA packets have a 12-byte header (length, clientID, frame), which is *per sender*. Each OpenRA packet is also sent on the TCP socket with a single send call, and the TCP socket has NoDelay enabled, which means it also incurs the 40-byte TCP header cost (best case is 20 for IP and 20 for TCP, could be higher with options). Each player sends orders every frame, whether or not they actually did anything because lockstep requires every player receives exact orders from every other player before progressing. This means that each player receives 7 order packets (they don't see their own) every net tick. 7 packets * 8 net ticks * 52 bytes = 2912 bytes/s in pure order overhead (assuming no packet loss).

Every player also sends a sync hash every net frame, and sync hash orders are [13 bytes long](https://github.com/OpenRA/OpenRA/blob/8c9b9df125a047b8d0bdd4c34ab0f221dc82d73f/OpenRA.Game/Network/Order.cs#L53) (includes order type byte), so these packets are 13 + 52 = 65 bytes long. 7 packets * 8 net ticks * 65 bytes = 3640 bytes/s. (Please note that this is also completely redundant since the server checks the sync hashes itself, it only forwards them for simplicity.) So far we've got 6552 bytes/s, which is ~52kbps (more because we rounded down on the number of net ticks per frame). Not great, not terrible, although it will scale proportionally to game speed, so double for the *real* mode, fastest, that apparently nobody plays.

Except that this is all the data required to say "nothing happened, everything is good". We could probably encode all of that data per net tick into less than the 12-byte difference between a single UDP packet and a TCP packet, including custom ack protocol data (4-byte bitfield for redundant ack status of last 32 frames), 1-byte bitfield for clients with orders, 4-byte frame number). We'd have so much to spare that we could up net ticks to every frame, which would actually reduce latency noticeably (to consistently a single game frame if your ping is sub 40ms, which is common nowadays).

Now, orders themselves are currently not very compressed. The order types are string encoded, but probably the worst possible offender is group orders. Each unit in the group is included in the order, and each unit has a 4-byte ID. Let's say we have 8 players with armies of 100 units each move clicking once per second. That's at least 4 bytes per unit * 100 units * 7 players = 2800 bytes/s of just unit IDs! Wait, that's actually not that much... If they were all spam clicking 5 times per second at the same time, it would start to get excessive, sure, but I don't think most players in this game have that high of an APM. It's also possible that they build way more than 100 units each, but seems unlikely to me. There's more data in these orders, such as order strings, target strings, locations, a 1-byte order type, but group move orders are the only type I can think of that would be common enough and large enough to be truly excessive, and the rest of the order will be fairly small in size.

I think this shows just how utterly insane unchecked network overhead could be, we should do some measurements and see if the napkin math checks out.

(Also, the funny thing is that the additional network overhead of my acks is probably greater than the amount saved by emptying those acks instead of just letting the server loopback, although it is still a save since both have that same overhead and we need one of them.)

jrb0001 reviewed (commented):
This isn't a code review, I just focused on the parts that looked relevant to the network protocol.

jrb0001 commented:
@matjaeck I am not involved in any mod decisions.
SP: @ABrandau, @ZxGanon
RV: @MustaphaTR , @ZxGanon
AS (Engine): @GraionDilach

I maintain a custom server implementation used for the official SP and RV servers. I definitely won't start working on it before it is on bleed and all backwards/forwards compatibility regressions got fixed. After that it will depend mostly on the needs of SP and RV.

adammitchelldev commented:
Just had a thought on this, with the server-side sync, we could change order sending to dispatch each order immediately as it is made rather than collecting them, removing the latency between giving the order and the next net tick (currently up to 120ms).

adammitchelldev commented:
Added buffering for frame data and sync packets from the server (it would be possible to buffer everything but the Server code is unwieldy). This prevents there being a ton of TCP/IP header overhead in bigger games due to the behavior of `NoDelay` as each order packet and sync packet from each player were being sent in separate TCP packets for the same frame. This also means that the 3:1 net-tick ratio is no longer "necessary", so I have removed it, although it does also increase the Sync Hash frequency (can add an independent ratio for that if desired). A more ideal solution down the line is to use a reliable UDP library or implement a custom UDP protocol since TCP will still suffer latency issues under poor network conditions even with `NoDelay` enabled.

I'd really encourage people to build and try this out, it's a **very** noticeable difference even without further optimizations on the game loop. Buildings are placed instantly and units shoot instantly. If I have to, I'll add a setting to switch between this and a simulated "old mode" just to hammer the difference home.

This PR effectively cuts minimum input latency in multiplayer from 400ms to 40ms _and_ reduces the impact of low latency players on game quality.

adammitchelldev committed:
Add new low-latency server-sync netcode mode

Co-authored-by: Leonardo <index.xp@gmail.com>

adammitchelldev committed:
Add netcode option to advanced tab

adammitchelldev commented:
Cleaned up to the point that it could be genuinely tested. Some known "issues" that could be improved:

- Singleplayer doesn't seem to respect the `UseNewNetcode` option when enabled.
- Multiplayer gameplay is not as smooth as it could be visually at high game speeds; the catchup logic is aggressive and drops render frames too eagerly. Solutions:
  - Only drop render frames if the catchup is significant.
  - Add a "jitter buffer" to prevent network jitter from causing framerate jitter. 
- Sometimes the game can slow down significantly because the server slowdown logic is not very scientific, it works a bit like a backseat driver screaming "SLOW DOWN" and the server panicking. ~~It's most often noticeable if a player is slow to load at the start of a high-speed game.~~ EDIT: This case is fixed now. The slowdown algorithm is still not tuned for actual slowdowns later in the game but it has been tested.

pchote commented:
> It's most often noticeable if a player is slow to load at the start of a high-speed game.

This definitely needs to be accounted for / fixed, as improving the behaviour for mods with long load times is about to become a major focus thanks to the remastered assets.

adammitchelldev commented:
I can add a wait in the server for the first frame of orders to arrive from everyone before ticking off.

adammitchelldev committed:
Add loading state to server

adammitchelldev commented:
Added a loading state so that the server does not start ticking until all players have specifically said they are ready, however, players leave the loading screen when they have finished loading, not when the game has started, so the current behavior is that the player is greeted with fog of war until the game actually starts. Changing this is probably going to be a bit hacky and ~~I'll add it later~~. Right now this is limited to new netcode to maintain the backwards compatibility but I imagine it'll be the right way to do things by default.

pchote commented:
The starting fog/shroud is a known issue and already on my radar to fix as part of the remaster work. Out of scope for here!

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -68,15 +78,36 @@ public void StartGame()
+			if (LobbyInfo.Clients.Count == 0)
+			{
+				frameData.AddClient(Connection.LocalClientId);
+			}
```
remove `{ }`

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -68,15 +78,36 @@ public void StartGame()
+			}
+
+			foreach (var client in LobbyInfo.Clients)
+			{
+				if (!client.IsBot)
+					frameData.AddClient(client.Index);
+			}
+
```
Here too!

adammitchelldev replied:
Are you sure? I get a lint error with VSC for removing these `Add braces to 'foreach' statement. [OpenRA.Game]csharp(IDE0011)`

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -68,15 +78,36 @@ public void StartGame()
+			else
+			{
+				for (var i = 0; i <= OrderLatency; ++i)
+					SendOrders();
```
And here too

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -170,59 +203,151 @@ void CheckSync(byte[] packet)
+		{
+			var orders = frameData.OrdersForFrame(World).ToList();
+			foreach (var order in orders)
+			{
 				UnitOrders.ProcessOrder(this, World, order.Client, order.Order);
+			}
```
And here

adammitchelldev replied:
I get the same/similar lint errors here: it doesn't like having the braces off of the else (due to multi-level) and it doesn't like a mismatch between the if and the else.

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -170,59 +203,151 @@ void CheckSync(byte[] packet)
+			if (Connection is NetworkConnection c)
+			{
+				c.UseNewNetcode = LobbyInfo.GlobalSettings.UseNewNetcode;
+			}
```
!

teinarss commented on OpenRA.Game/Network/ReplayConnection.cs:
```diff
@@ -124,35 +124,49 @@ public ReplayConnection(string replayFilename)
-				packetFn(LocalClientId, sync[0]);
-				sync.RemoveAt(0);
+				packetFn(LocalClientId, sync.Dequeue());
 			}
```
!

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -259,25 +240,37 @@ public Server(List<IPEndPoint> endpoints, ServerSettings settings, ModData modDa
+					if (LobbyInfo.GlobalSettings.UseNewNetcode && State == ServerState.GameStarted)
 					{
-						EndGame();
-						break;
+						localTimeout = serverGame.MillisToNextNetFrame * 1000;
+					}
+					else
+					{
+						// Block for at most 1 second in order to guarantee a minimum tick rate for ServerTraits
+						// Decrease this to 100ms to improve responsiveness if we are waiting for an authentication query
+						localTimeout = waitingForAuthenticationCallback > 0 ? 100000 : 1000000;
 					}
```
!

adammitchelldev replied:
It _really_ doesn't like having comments without the braces, shows up as an error, not just a lint warning for me.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -805,17 +810,76 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+					GameSave.DispatchOrders(conn, frame, data);
+			}
+			else if (conn != null && data.Length != 0 && (OrderType)data[0] == OrderType.SyncHash)
+			{
+				// TODO: Find a less hacky way to deal with synchash relaying
+				DispatchOrdersToOtherClients(conn, frame, data, true);
+			}
+			else if (conn != null && LobbyInfo.GlobalSettings.UseNewNetcode && State >= ServerState.GameLoading)
+			{
+				serverGame.OrderBuffer.BufferOrders(conn.PlayerIndex, data);
+			}
 			else
-				DispatchOrdersToClients(conn, frame, data);
+			{
+				// TODO: Remove frame-based order relaying unless game is running
+				DispatchOrdersToOtherClients(conn, frame, data);
+			}
```
!

adammitchelldev replied:
Lint errors here too: I get `Add braces to 'else' statement. [OpenRA.Game]csharp(IDE0011)` and `Single-line comment should be preceded by blank line [OpenRA.Game]csharp(SA1515)`.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -805,17 +810,76 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+		public void DispatchBufferedOrdersToOtherClients(int fromClient, List<byte[]> allData)
+		{
+			var ms = new MemoryStream(allData.Select(d => d.Length).Sum());
+			foreach (var data in allData)
+			{
+				ms.WriteArray(data);
+			}
```
!

teinarss replied:
Still here :)

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1031,6 +1096,35 @@ void InterpretServerOrder(Connection conn, Order o)
+								var amount = int.Parse(o.TargetString);
+								serverGame.SlowDown(amount);
+							}
+							else
+							{
+								Log.Write("server", "Received request to slow down from client {0} but not using new netcode!", conn.PlayerIndex);
+							}
```
!

adammitchelldev replied:
It doesn't like mismatched else but it's technically not an error so I can change it.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1031,6 +1096,35 @@ void InterpretServerOrder(Connection conn, Order o)
+							else
+							{
+								Log.Write("server", "Received Loaded message from client {0} but not using new netcode!", conn.PlayerIndex);
+							}
```
!

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1240,17 +1338,30 @@ public void StartGame()
+				else
+				{
+					State = ServerState.GameStarted;
+				}
```
!

pchote commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -29,8 +29,9 @@ namespace OpenRA.Server
 		WaitingPlayers = 1,
-		GameStarted = 2,
-		ShuttingDown = 3
+		GameLoading = 2,
```
Can we please track this using a bool inside the `GameStarted` state instead?

This enum is part of the master server API, so changing these values would require a new protocol version and changes to the masterserver and online server list. A big task for a small feature.

In the future I want to allow clients to share their current loading percentage, so that bool would be replaced by a dictionary of client to number.

adammitchelldev replied:
Sure thing

adammitchelldev committed:
Remove braces for single-line consequents

adammitchelldev committed:
Use a separate GameState control in server

adammitchelldev committed:
Add smoothing to catchup frames

anvilvapre commented:
Related #19019.

teinarss commented:
Tested launching a server with the UseNewNetcode set to true, after joining with the first client the server stopped responding to new orders.

adammitchelldev commented:
Will check that out. Was this a dedicated server? It has been working for
me for non-dedicated.

On Wed, 13 Jan 2021, 22:09 teinarss, <notifications@github.com> wrote:

> Tested launching a server with the UseNewNetcode set to true, after
> joining with the first client the server stopped responding to new orders.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/18783#issuecomment-759770099>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOKXUDX76XUGALQBTEGZ3TSZYK2DANCNFSM4TC5KOYQ>
> .
>

teinarss commented:
Yes it was a dedicated server.

On Thu, 14 Jan 2021, 01:05 Adam Mitchell, <notifications@github.com> wrote:

> Will check that out. Was this a dedicated server? It has been working for
> me for non-dedicated.
>
> On Wed, 13 Jan 2021, 22:09 teinarss, <notifications@github.com> wrote:
>
> > Tested launching a server with the UseNewNetcode set to true, after
> > joining with the first client the server stopped responding to new
> orders.
> >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/OpenRA/OpenRA/pull/18783#issuecomment-759770099>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAOKXUDX76XUGALQBTEGZ3TSZYK2DANCNFSM4TC5KOYQ
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/18783#issuecomment-759835969>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESI7GAVSUR5MF66V5XSW4TSZYYKZANCNFSM4TC5KOYQ>
> .
>

adammitchelldev committed:
Improve smoothness in tickrate

adammitchelldev commented:
Hmm, I'm not getting the same problem locally. If it's still happening for you after building the latest commit then I'll need you to help investigate it for me sorry! Are you on windows?

adammitchelldev commented:
I took some videos of the new netcode to compare to the old one. Sadly I'm stuck on a pretty old macbook at the moment so the video framerate is choppy, but it's still a noticeable difference.

[Old](https://youtu.be/xOHVcK9hx-M)

[New](https://youtu.be/GGCkfC-yGEA)

adammitchelldev commented:
I will also be hosting a "New Netcode" and "Old Netcode" dedicated server on GCP (London) for a little while for demonstration purposes, they should appear in the server browser if you are playing on `{{DEV_VERSION}}`. Please be aware that the old netcode cheats when there is only one player to reduce latency, you need to connect with two clients to get the proper experience!

deleted-user-1 commented:
> I will also be hosting a "New Netcode" and "Old Netcode" dedicated server on GCP (London) for a little while for demonstration purposes, they should appear in the server browser if you are playing on `{{DEV_VERSION}}`. Please be aware that the old netcode cheats when there is only one player to reduce latency, you need to connect with two clients to get the proper experience!

It think it makes sense to give it a version so that testers know which built/commit they need. I think it should be enough to run `make version` on linux and it will insert the commit hash, alternatively insert it manually with `make version VERSION=<hash/date/whateveryoulike>`. Please correct me if that is wrong.

adammitchelldev commented:
My experience with `make version` prior was that it does not always produce the same version for everyone due to forks having potential short hash conflicts, so I will re-instate the method I had earlier.

teinarss commented on OpenRA.Game/Server/ServerGame.cs:
```diff
@@ -0,0 +1,84 @@
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
```
can be removed.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -808,17 +824,72 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
 			if (frame == 0 && conn != null)
+			{
 				InterpretServerOrders(conn, data);
+				if (GameSave != null)
+					GameSave.DispatchOrders(conn, frame, data);
```
`GameSave?.DispatchOrders(conn, frame, data);`

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -808,17 +824,72 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+				ms.WriteArray(data);
+			}
+
+			Connection conn = Conns.FirstOrDefault(c => c.PlayerIndex == fromClient);
```
use `var`

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -808,17 +824,72 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+			writer.Write((byte)OrderType.Ack);
+			writer.Write((short)acks);
+
+			Connection conn = Conns.FirstOrDefault(c => c.PlayerIndex == forClient);
```
Same  here.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -890,11 +961,12 @@ void InterpretServerOrder(Connection conn, Order o)
-							if (!OpenRA.Exts.TryParseInt64Invariant(o.TargetString, out var pingSent))
+							long pingSent;
+							if (!OpenRA.Exts.TryParseInt64Invariant(o.TargetString, out pingSent))
 							{
```
`out var pingSent`

adammitchelldev replied:
Must have been a faulty merge

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1051,58 +1146,62 @@ public void DropClient(Connection toDrop)
-					}
 
-					DispatchOrders(toDrop, toDrop.MostRecentFrame, new[] { (byte)OrderType.Disconnect });
+						if (serverGame != null)
+							serverGame.OrderBuffer.DropClient(toDrop.PlayerIndex);
```
`serverGame?.OrderBuffer.DropClient(toDrop.PlayerIndex);`

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1243,17 +1342,32 @@ public void StartGame()
+					GameState = GameState.Playing;
+				}
+
+				// TODO remove: this "pre-disconnecting" method adds unnecessary complexity just for the sake of the replay stream
+				/*var disconnectData = new[] { (byte)OrderType.Disconnect };
```
Do we want to remove this code?

adammitchelldev replied:
I'm not sure yet, I haven't made replay and savegame testing a priority and I think that's where it was required. I like to think I actually fixed that one already.

adammitchelldev replied:
I'll remove it

teinarss commented on OpenRA.Game/Server/OrderBuffer.cs:
```diff
@@ -0,0 +1,84 @@
+ */
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
```
using System;
using System.Collections.Generic;

teinarss commented on OpenRA.Game/Server/Connection.cs:
```diff
@@ -120,6 +122,59 @@ public void ReadData(Server server)
+			}
+		}
+
+		private void InnerSend(byte[] data)
```
remove `private`

adammitchelldev replied:
God damn fucking Java

teinarss commented on OpenRA.Game/Network/ReplayConnection.cs:
```diff
@@ -124,35 +124,47 @@ public ReplayConnection(string replayFilename)
+			}
+		}
+
+		public int LastAckedFrame
+		{
+			get { return ordersFrame;  }
 		}
```
Not used.

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -789,13 +803,15 @@ void HandleSyncOrder(int frame, byte[] packet)
-				DispatchFrameToClient(c, from, frameData);
+				DispatchFrameToClient(c, frameData, buffer);
 
+			/*
```
This needs to be added back

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -68,15 +79,28 @@ public void StartGame()
-			if (GameSaveLastFrame < 0)
-				for (var i = NetFrameNumber; i <= FramesAhead; i++)
-					Connection.Send(i, new List<byte[]>());
+			// HACK: FramesAhead is only ever 0 in singleplayer, so we increase the rate of apparent net ticks to decrease latency
+			// if (OrderLatency == 0)
+			// 	NetTickScale = 1;
```
Remove?

teinarss commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -9,6 +9,7 @@
  */
 #endregion
 
+using System;
```
Can be removed.

teinarss commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +28,59 @@ public override string ToString()
-			clientQuitTimes[clientId] = lastClientFrame;
+		public void ClientQuit(int clientId)
+		{
+			// TODO review whether we need to specify quit frame
```
A wild todo appears!

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,120 @@ void NetworkConnectionReceive()
+				byte[] queuedPacket = default;
+				if (awaitingAckPackets.Count > 0 && !awaitingAckPackets.TryDequeue(out queuedPacket))
+				{
+					// The dequeing failed because of concurrency, so we retry
```
typo dequeuing

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,120 @@ void NetworkConnectionReceive()
+				ms.WriteArray(BitConverter.GetBytes(frame));
+				ms.WriteArray(syncData);
+
+				queuedSyncPackets.Add(ms.GetBuffer()); // TODO: re-add sync packets
```
A todo :)

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,120 @@ void NetworkConnectionReceive()
+				byte[] ackArray;
+				using (var ackMs = new MemoryStream(ordersLength))
+				{
+					foreach (var o in orders)
+					{
+						ackMs.WriteArray(o);
+					}
```
remove { }

teinarss commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,120 @@ void NetworkConnectionReceive()
-					ms.WriteArray(BitConverter.GetBytes(q.Length));
-					ms.WriteArray(q);
-					base.Send(q);
+					awaitingAckPackets.Enqueue(ackArray); // TODO fix having to write byte buffer twice
```
Todo!

teinarss commented on OpenRA.Game/Game.cs:
```diff
@@ -594,55 +595,64 @@ static void InnerLogicTick(OrderManager orderManager)
-					if (world == null)
-						return;
-
-					var isNetTick = LocalTick % NetTickScale == 0;
+				return;
+			}
```
Remove { }

teinarss commented on OpenRA.Game/Game.cs:
```diff
@@ -823,7 +840,8 @@ static void Loop()
 
 						// Force at least one render per tick during regular gameplay
-						if (OrderManager.World != null && !OrderManager.World.IsLoadingGameSave && !OrderManager.World.IsReplay)
+						// TODO: Have OrderManager tell us when we should force render
```
!

adammitchelldev replied:
I intended this to mean TODO in a later PR, removing

adammitchelldev commented:
I've created a preview release, which people can try out here:

https://github.com/adammitchelldev/OpenRA/releases/tag/devtest-netcode-1

Testimonials: "this is a game changer" - @Mesacer

I have a couple of dedicated servers for both Red Alert and Tiberian Dawn open and ready for anyone brave enough to test it, plus a server with the old netcode to compare (I can expand the dedicated servers on request).

adammitchelldev committed:
Address review cleanup requests

adammitchelldev commented:
Noted issues (thanks to bug):

1. Buildings seem to explode too early (needs investigation).
2. Rally point still has some delay (I suspect this is not related to these changes).
3. Replays crash (oops).

adammitchelldev committed:
Fix possible negative timeout on socket select

anvilvapre commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,118 @@ void NetworkConnectionReceive()
+				if (awaitingAckPackets.Count > 0 && !awaitingAckPackets.TryDequeue(out queuedPacket))
+				{
+					// The dequeuing failed because of concurrency, so we retry
+					for (var c = 0; c < 5; c++)
```
Seems buggy. Suspicious. Active for loop hoping a lock will become freed.

adammitchelldev replied:
This was not my code actually, originally written by @Leonardo-Ferreira, I didn't have the loop prior but I assumed he added it for a good reason.

anvilvapre commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,118 @@ void NetworkConnectionReceive()
 
-			try
+			// Always write data for old netcode
+			if (orders.Count() > 0 || !UseNewNetcode)
```
Avoid LINQ. Walking a complete list.

teinarss replied:
In this case it should be fine, the underlying structure is a list. But the parameter `orders` could be kept as a list or changed to an array.

pchote replied:
I agree with @teinarss

anvilvapre commented:
Can you briefly please explain what/how you actually implemented what you aimed for? It helps to understand the code.
- You do no longer send SyncFrames after an empty frame? A frame with no orders. You queue them.
- You send the sync frames at/with/after the next frame that does have at least one Order?
- Because of that you need to cache the empty frames for which you didn't receive a sync frame after that empty yet?
- Because you only send the sync for multiple frames if first frames are empty there is no chance of out-of-sync?
- It won't cause the game to crash at the frame it becomes out-of-sync and for which it it didn't receive a sync yet?

I understand your intent, but now how you actually implemented it, since it are a lot of scattered changes. Also because I have limited knowledge of the protocol, but i assume many.

So all clients know they are `talking` the new protocol - over the old protocol by checking the lobby settings - which are synced? So only based a client will know and will not crash? So the server is unware of the protocol changes, but all clients do need to know?

In a future version you could perhaps calculate the hash across multiple sync frames and send that as a master sync for multiple frames - with the advantage it would be smaller in size?

pchote commented:
> You do no longer send SyncFrames after an empty frame? A frame with no orders. You queue them.

This is potentially a problem if true. It is important to catch the desync as soon as it happens - if the game state advances beyond the history buffer that we track then we will lose the ability to debug them.

anvilvapre commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +334,118 @@ void NetworkConnectionReceive()
 
-		protected override void Send(byte[] packet)
+		// Override send frame orders so we can hold them until ACK'ed
+		public override void Send(int frame, IEnumerable<byte[]> orders)
```
Wasn't `base.Send` needed for Replay?

adammitchelldev replied:
I will investigate

teinarss replied:
@adammitchelldev just looking at the code im guessing the answer to the question above should be no,  right?

adammitchelldev commented:
> * You do no longer send SyncFrames after an empty frame? A frame with no orders. You queue them.

> This is potentially a problem if true. It is important to catch the desync as soon as it happens - if the game state advances beyond the history buffer that we track then we will lose the ability to debug them.

Yes, this may be an issue. The idea is to cut down network traffic up, but it shouldn't be hard to fix. This is only a significant issue when sync hash reports are enabled right? I may fix it by turning on immediate sync hash sending if sync hash reports are enabled, being able to turn it off is a big save.

> * You send the sync frames at/with/after the next frame that does have at least one Order?

Yes, I think, this can be changed.

> * Because of that you need to cache the empty frames for which you didn't receive a sync frame after that empty yet?

Yes, this was already the case.

> * Because you only send the sync for multiple frames if first frames are empty there is no chance of out-of-sync?

Not sure I understand this question.

> * It won't cause the game to crash at the frame it becomes out-of-sync and for which it it didn't receive a sync yet?

No, I don't think so, but this was already the case right? The game can't know it's out of sync until it receives a sync hash telling it so.

> I understand your intent, but now how you actually implemented it, since it are a lot of scattered changes. Also because I have limited knowledge of the protocol, but i assume many.
> 
> So all clients know they are `talking` the new protocol - over the old protocol by checking the lobby settings - which are synced? So only based a client will know and will not crash? So the server is unware of the protocol changes, but all clients do need to know?

Yes, the clients need to know which protocol to use. The reason it is optional on the server is twofold:

* If instabilities are discovered, we want a way to quickly disable the netcode change and continue playtesting a build.
* Third party server implementations may not want to immediately update to the new netcode, so they can simply disable the new setting and use the old netcode without much or any modification.

> In a future version you could perhaps calculate the hash across multiple sync frames and send that as a master sync for multiple frames - with the advantage it would be smaller in size?

The hash size is not a problem, it's only a few bytes, the balance is more on the size of the sync hash reports which are kept locally, since it needs to keep enough reports buffered to match the last sync hash coming in. The "rolling" sync technique would be better just for detecting desync "soon enough", but not for trying to debug it.

anvilvapre commented:
> > * Because you only send the sync for multiple frames if first frames are empty there is no chance of out-of-sync?
> 
> Not sure I understand this question.

It was similar to the question after. I thought because the frame(s) (always max 1?) was empty that perhaps the sync hash also wouldn't change. But that is not correct, other client orders are received, but also the tick time in probably part of the sync hash. So in summary - there does seem to be more chance of out undetected out of sync. 

The sync hash these days also contains the player state. Will it mean it will be synced later also?

How will you recover from lost or out of order packets in future when changing to UDP?

adammitchelldev commented:
> The sync hash these days also contains the player state. Will it mean it will be synced later also?

The sync hash should depend on all game state, so anything different should cause a desync. I looked at the code and sync hash packets are still send even if there are no orders. See `Connection.Send`, at the bottom, we always dequeue and send sync hash packets.

> How will you recover from lost or out of order packets in future when changing to UDP?

Using techniques to create reliability over UDP, such as those described by https://gafferongames.com/post/reliable_ordered_messages/ plus some redundancy adapted to meet bandwidth expectations.

The major benefit is that these techniques can be optimized for latency, which is where TCP really suffers. A missed packet in TCP can cause an additional **2** round trips of latency, plus a 200ms exponentially increasing "backoff". However, it's not a big enough benefit to worry about right this second, not in comparison to these changes.

adammitchelldev committed:
Implement new slowdown and catchup algorithm

jrb0001 commented:
> Yes, this may be an issue. The idea is to cut down network traffic up, but it shouldn't be hard to fix. This is only a significant issue when sync hash reports are enabled right? I may fix it by turning on immediate sync hash sending if sync hash reports are enabled, being able to turn it off is a big save.

If you want to reduce network traffic, then you should optimize the framing and order format. The sync hash is very cheap, especially when you can batch it together with normal orders. Worst case with 50 ticks per second and no other activity at all is 50 * (1 + 4 + 4 + 4 + 8 + 20 + 40) = 4050 bytes per second. That's nothing and the overhead decreases when you can group it together with orders. You shouldn't optimize the idle bandwith usage, you should focus on the peaks in my opinion.

> The major benefit is that these techniques can be optimized for latency, which is where TCP really suffers. A missed packet in TCP can cause an additional **2** round trips of latency, plus a 200ms exponentially increasing "backoff". However, it's not a big enough benefit to worry about right this second, not in comparison to these changes.

Building your own thing ontop of UDP **can** give you better latency, but you will handle all edge cases yourself or you might end up with worse latency, broken connections or even collateral damage to other services/users. I would ignore it at least for now, there are easier and more effective changes to be done, like some in this PR.

adammitchelldev commented:
> > Yes, this may be an issue. The idea is to cut down network traffic up, but it shouldn't be hard to fix. This is only a significant issue when sync hash reports are enabled right? I may fix it by turning on immediate sync hash sending if sync hash reports are enabled, being able to turn it off is a big save.

It's actually sending every frame anyway, it sends even when there are no orders, it's only "queued" so that it is sent within the order packet to avoid doubling the base TCP overhead.

> If you want to reduce network traffic, then you should optimize the framing and order format. The sync hash is very cheap, especially when you can batch it together with normal orders. Worst case with 50 ticks per second and no other activity at all is 50 * (1 + 4 + 4 + 4 + 8 + 20 + 40) = 4050 bytes per second. That's nothing and the overhead decreases when you can group it together with orders. You shouldn't optimize the idle bandwith usage, you should focus on the peaks in my opinion.

I agree, I don't think it needs optimization further than what I've done here already. It was the *down* bandwidth that was bad because each player order and hash sync was sent as a separate packet by the server every frame. That's fixed in this PR though.

adammitchelldev committed:
Add network performance measurements

adammitchelldev committed:
Fix replays

adammitchelldev committed:
Fix replays on old netcode

teinarss commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,87 @@ public override string ToString()
+		public bool TryPeekTimestep(out int timestep)
 		{
-			return !ClientsNotReadyForFrame(frame).Any();
+			if (timestepData.Any())
+			{
+				timestep = timestepData.Dequeue();
+				return true;
+			}
```
```suggestion
			if (timestepData.TryDequeue(out timestep))
			{
				return true;
			}
			else
			{
				timestep = 0;
				return false;
			}
```

abcdefg30 replied:
Style nit: No braces around the first if.

teinarss commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,87 @@ public override string ToString()
-			var frameData = framePackets.GetOrAdd(frame);
-			return ClientsPlayingInFrame(frame)
-				.Where(client => !frameData.ContainsKey(client));
+			if (timestepData.Any())
+				timestepData.Dequeue();
```
```suggestion
			timestepData.TryDequeue(out _);
```

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -170,59 +226,180 @@ void CheckSync(byte[] packet)
 			{
-				return GameStarted
-					? frameData.ClientsNotReadyForFrame(NetFrameNumber)
+				return GameStarted // TODO review, compare to bleed
```
!

teinarss commented on OpenRA.Game/Network/ReplayConnection.cs:
```diff
@@ -110,8 +109,9 @@ public ReplayConnection(string replayFilename)
 						// Don't replace the final disconnection packet - we still want this to end the replay.
-						if (client == lastClientToDisconnect)
-							continue;
+						// TODO Ensure that this is not necessary
+						/*if (client == lastClientToDisconnect)
+							continue;*/
```
!

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -808,17 +825,69 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
 				}
 			}
+
+			// TODO: Make this nicer
```
!

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -808,17 +825,69 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+			else if (conn != null && LobbyInfo.GlobalSettings.UseNewNetcode && State >= ServerState.GameStarted)
+				serverGame.OrderBuffer.BufferOrders(conn.PlayerIndex, data);
+
+			// TODO: Remove frame-based order relaying unless game is running
```
!

teinarss commented on OpenRA.Game/Server/ServerGame.cs:
```diff
@@ -0,0 +1,84 @@
+		public ServerGame(int worldTimeStep)
+		{
+			CurrentNetFrame = 1;
+			NetTimestep = worldTimeStep * Game.NewNetcodeNetTickScale; // TODO: Set net tick scale via lobby settings
```
!

adammitchelldev committed:
Cleanup and address PR comments

teinarss commented on OpenRA.Game/Game.cs:
```diff
@@ -791,16 +796,16 @@ static void Loop()
 				// Ideal time between logic updates. Timestep = 0 means the game is paused
 				// but we still call LogicTick() because it handles pausing internally.
-				var logicInterval = worldRenderer != null && worldRenderer.World.Timestep != 0 ? worldRenderer.World.Timestep : Timestep;
+				// var logicInterval = worldRenderer != null && worldRenderer.World.Timestep != 0 ? worldRenderer.World.Timestep : OrderManager.SuggestedTimestep;
```
remove?

teinarss commented on OpenRA.Game/Game.cs:
```diff
@@ -812,18 +817,20 @@ static void Loop()
 					if (now >= nextLogic && !renderBeforeNextTick)
 					{
-						nextLogic += logicInterval;
+						// nextLogic += logicInterval;
```
Here too?

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -35,11 +40,44 @@ public sealed class OrderManager : IDisposable
 		public int LocalFrameNumber;
-		public int FramesAhead = 0;
 
+		public int SyncFrameScale = 1; // TODO make this based on actual time, I would suggest once per second
```
Something for a followup?

adammitchelldev replied:
Same as the other sync hash thing

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -35,11 +40,44 @@ public sealed class OrderManager : IDisposable
+			}
+		}
+
+		// int suggestedCatchupTimestep;
```
Rewove?

teinarss commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -68,15 +107,30 @@ public void StartGame()
-			if (GameSaveLastFrame < 0)
-				for (var i = NetFrameNumber; i <= FramesAhead; i++)
-					Connection.Send(i, new List<byte[]>());
+			// suggestedCatchupTimestep = World.Timestep;
```
Same here?

teinarss commented on OpenRA.Game/Network/ReplayConnection.cs:
```diff
@@ -123,36 +118,41 @@ public ReplayConnection(string replayFilename)
+		public void Send(int frame, IEnumerable<byte[]> orders) { }
 		public void SendImmediate(IEnumerable<byte[]> orders) { }
 
+		// TODO: Fix this HACK
+		// so that replays are not dependent on SendSync for their timing and we can optionally reduce Sync rate
```
For a followup pr?

adammitchelldev replied:
Yes

teinarss commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -805,17 +829,67 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
+			var from = fakeFrom != null ? fakeFrom.PlayerIndex : 0;
+
+			var frameData = CreateFrame(from, frame, data);
+			foreach (var c in Conns.ToList())
```
No need to use `.ToList()`

reaperrr commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,77 @@ public override string ToString()
+		{
+			return ClientsPlayingInFrame()
+				.SelectMany(x => framePackets[x].Dequeue().ToOrderList(world)
+					.Select(y => new ClientOrder { Client = x, Order = y }));
```
Proposed alternative based on #19011:
```
			// PERF: Avoid LINQ
			foreach (var clientId in ClientsPlayingInFrame())
			{
				// Assumes IsReadyForFrame is true --> check if correct before we merge this
				var frameData = framePackets[clientId];
				var orderList = OrderIO.ToOrderList(frameData.Dequeue(), world);
				foreach (var order in orderList)
					yield return new ClientOrder { Client = clientId, Order = order };
			}
```

According to my Stopwatch testing, this reduces (Stopwatch/system, not game) ticks by approximately 1/3 on average for populating `var orders` in `OrderManager.ProcessOrders()`.

On my system, that means a reduction from ~27K stopwatch/sys ticks on first frame of RA shellmap to ~17.7K. The absolute wins on later frames are much lower of course, but it still goes from an average of 18ish to 12ish per order frame.

adammitchelldev replied:
I'm happy to merge this change because it is cool but this code is not called frequently.

reaperrr replied:
I know, but I think at least the reduction at game launch for busy scripted maps like the RA shellmap makes this worthwhile.

reaperrr commented:
This currently breaks the FrameLimiter (checking it and setting something like 60 has no effect on my system).

reaperrr commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,77 @@ public override string ToString()
+		public bool IsReadyForFrame()
+		{
+			return !ClientsNotReadyForFrame().Any();
+		}
```
Also based on #19011:
```
			// PERF: Avoid LINQ
			foreach (var client in ClientsNotReadyForFrame())
				return false;

			return true;
```

together with...

reaperrr replied:
See comment below.

reaperrr commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,77 @@ public override string ToString()
+		{
+			return ClientsPlayingInFrame()
+				.Where(client => framePackets[client].Count == 0);
+		}
```
```
			// PERF: Avoid LINQ
			foreach (var client in ClientsPlayingInFrame())
				if (framePackets[client].Count == 0)
					yield return client;
```

~yielded roughly a 30% reduction in stopwatch/sys ticks for the `IsReadyForFrame` check in `OrderManager.TryTick`.~

~Not much in absolute terms, but more than nothing.~
~I measured about 2 stopwatch ticks less on RA's shellmap and and 3 less on a large map with me and 9 bots.~

On the RA shellmap at frame 0 this actually reproduceably regresses performance a bit, much less than the suggested `OrdersForFrame` changes win us but probably still too much to make the miniscule wins on regular ticks worth it.

adammitchelldev commented:
> This currently breaks the FrameLimiter (checking it and setting something like 60 has no effect on my system).

Sounds likely, there are some changes to the way the game loop works.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -31,10 +31,13 @@ namespace OpenRA
 		public const int Timestep = 40;
 		public const int TimestepJankThreshold = 250; // Don't catch up for delays larger than 250ms
-
+		public const double NetCatchupFactor = 0.2;
```
Can we please have a comment explaining what this is and the significance of 0.2?

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -309,9 +313,15 @@ void NetworkConnectionReceive()
-					if (len == 0)
+
+					if (UseNewNetcode && client == LocalClientId && len == 9 && buf[4] == (byte)OrderType.Ack)
+					{
```
Style nit: omit braces on single statements

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +335,102 @@ void NetworkConnectionReceive()
 
+		void Ack(byte[] buf)
+		{
+			var reader = new BinaryReader(new MemoryStream(buf));
```
Style nit: avoid using `BinaryReader`. We have our own stream extension methods for reading!

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +335,102 @@ void NetworkConnectionReceive()
+			for (var i = 0; i < framesToAck; i++)
+			{
+				if (!awaitingAckPackets.TryDequeue(out var queuedPacket))
+				{
```
braces

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -325,35 +335,102 @@ void NetworkConnectionReceive()
 
-				foreach (var q in queuedSyncPackets)
+				if (UseNewNetcode)
+				{
```
braces

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -362,6 +439,37 @@ protected override void Send(byte[] packet)
+			{
+				int listLengthNeeded = queuedSyncPackets.Sum(i => 4 + i.Length);
+				if (ms.Capacity - ms.Length < listLengthNeeded)
+				{
```
braces here and in the else

pchote commented on OpenRA.Game/Network/FrameData.cs:
```diff
@@ -27,55 +27,77 @@ public override string ToString()
+
+		public void AddFrameOrders(int clientId, byte[] orders, int timestep)
+		{
+			// HACK: Due to design we can actually receive client orders before the game start order
```
This sounds sketchy. Is this referring to the server- or client-side?

I'm pretty sure that by design non-immediate orders can't be sent before StartGame, so can't we just defer processing these until that has been processed?

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -19,6 +19,11 @@ namespace OpenRA.Network
 {
 	public sealed class OrderManager : IDisposable
 	{
+		const double CatchupFactor = 0.1;
```
Please add comments documenting the meaning and significance of these values!

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -19,6 +19,11 @@ namespace OpenRA.Network
 	{
+		const double CatchupFactor = 0.1;
+		const double CatchUpLimit = 1.5;
+		const int SimLagThreshold = 500;
```
Style nit: `SimLag` feels like it goes against our naming style, but I can't point out exactly why.

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -170,59 +226,180 @@ void CheckSync(byte[] packet)
 
-			foreach (var order in frameData.OrdersForFrame(World, NetFrameNumber))
+		/*
+		 * Only available if TickImmediate() is called first and we are ready to dispatch received orders locally.
```
Use `//` style comments.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -594,55 +597,57 @@ static void InnerLogicTick(OrderManager orderManager)
+			using (new PerfSample("tick_time"))
 			{
-				using (new PerfSample("tick_time"))
+				if (world == null || !orderManager.ShouldUseCatchUp || !orderManager.LobbyInfo.GlobalSettings.UseNewNetcode)
```
The `world == null` and `!orderManager.ShouldUseCatchUp` (which appears to be equivalent to `OrderManager.Connection != NetworkConnection`) cases look like they will be a problem here. What is our route to removing these once we want to remove the legacy netcode and tick logic in the future?

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -87,6 +141,10 @@ public OrderManager(ConnectionTarget endpoint, string password, IConnection conn
 			AddChatLine += CacheChatLine;
+
+			// HACK: should not rely on this
+			if (conn is NetworkConnection)
```
Can we define `ShouldUseCatchUp` as a property returning this instead?

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -35,11 +40,44 @@ public sealed class OrderManager : IDisposable
+		// int suggestedCatchupTimestep;
+		public volatile bool IsStalling;
 		public long LastTickTime = Game.RunTime;
+		public long RealTickTime = Game.RunTime;
```
I found these names quite confusing. Can we please use `TickStartRuntime` and `LastTickStartRuntime` to match the existing naming convention?

pchote commented on OpenRA.Game/Network/OrderManager.cs:
```diff
@@ -170,59 +226,180 @@ void CheckSync(byte[] packet)
 			++NetFrameNumber;
 		}
 
+		public void TickPreGame()
```
There are two things that don't sit well with me here:
1. The name describes the thing that calls this rather than what it actually does. We should rename this to something more descriptive or, better, fold it into a special case within `TryTick`.
2. I think that the logic here is also going to be needed for fixing the `ShouldUseCatchUp` issue i commented in Game above.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -594,55 +597,57 @@ static void InnerLogicTick(OrderManager orderManager)
-			var worldTimestep = world == null ? Timestep : world.IsLoadingGameSave ? 1 : world.Timestep;
+			var worldTimestep = orderManager.SuggestedTimestep;
+
 			var worldTickDelta = tick - orderManager.LastTickTime;
```
This appears to be the only place that uses `orderManager.LastTickTime` now, and this isn't updated reliably (only when `orderManager.TryTick` fails *and* `NetFrameNumber == 0`) so is basically a no-op in the new net code.

It looks like it should be possible to rewrite this using `orderManager.RealTickTime` &rarr; `TickStartRuntime` for both old and new netcodes, which would make this a lot clearer and hopefully less buggy (I imagine that the frame limiter bugs is also related).

pchote commented on OpenRA.Game/Network/Connection.cs:
```diff
@@ -153,27 +154,27 @@ protected virtual void Send(byte[] packet)
 			foreach (var p in packets)
 			{
-				packetFn(p.FromClient, p.Data);
+				packetFn(p.FromClient, p.Data, p.Timestep);
```
I really don't understand what is going on with passing this timestep through all the order plumbing. Why isn't this encoded inside  `p.Data` and dealt with at the higher level parsing?

pchote reviewed (commented):
I tried to make a start on reviewing this, but it is too big for me to wrap my head around in its current form, and as there are several potentially big issues to resolve that might require fundamental changes (fixing frame limiter, fixing desync handling, my comments about the game loop below) it feels like it would be better to wait for these fixes before continuing.

#19297 fixes a lot of the crazyness in the current timestep handling. Can we please rebase this on that PR and extract a new first commit that does the game loop changes, `orderManager.TryTick`, etc with the existing netcode, before the later commits add the new logic on top? This would give a better basis to continue reviewing from.

pchote commented:
To summarise with a single focused piece of feedback: I think that the next step forward here should be to find a way to reabstract the client side implementation such that the old/new netcode flag becomes an internal implementation detail of OrderManager and the main game loop behaves the same for both.

This will force us to confront the long standing design issues with the main loop, and I expect should also fix many of the issues raised above.

anvilvapre commented:
What is the status of this pull request?

pchote commented:
@teinarss has taken the lead on splitting this up and filing smaller focused PRs working towards the final goal. See #19374, #19402, #19447 for the parts we have merged so far.

abcdefg30 commented:
Closing this then as to-be-superseded.
